### DISCUSSION TITLE
Navigating from different thread

### DISCUSSION THREAD

**Discussion by NWHBakker** [zauberzeug/nicegui#3518] [Q&A] [answered]
### Question

Hello,

I'm trying to build an application with quite a bunch of callbacks from different threads, for which I sometimes need to open a different page. However, is mostly complains about creating UI from a background task.

`RuntimeError: The current slot cannot be determined because the slot stack for this task is empty.`

```py
from nicegui import ui
import threading
import time

ui.label("start_page")

def threaded_task():
    time.sleep(10)
    ui.navigate.to("/pagetwo")
    return

@ui.page("/pagetwo")
def second_page():
    ui.label("Second page")

task = threading.Thread(target=threaded_task)
task.start()
ui.run(show = False, reload=False)
```

Is there any way to do this? 

I've also tried using a ui.timer reading a global variable,which is being set in the thread. 
But this also doesn't work.

**Comment by rodja** [comment]
I suggest you rework your code to not use threads at all (or at least only for cpu intensive or io intensive work, see our [FAQ](https://github.com/zauberzeug/nicegui/wiki/FAQs#why-is-my-long-running-function-blocking-ui-updates)). Mixing threading with an event loop (asyncio) for concurrency is quite difficult. And NiceGUI builds upon FastAPI which is build around asnycio. If you must communicate between a background thread and the event loop, you may take inspiration from our progress bar example: https://github.com/zauberzeug/nicegui/blob/main/examples/progress/main.py.

**Reply by NWHBakker** [reply]
Yikes, reworking the code to not contain threads really isn't an option. 

It's a rather large system, and the ui was more of an afterthought. 
Because of that events aren't started from the ui, which makes the example you sent kinda unusable. 

Do you have any other suggestions? 
Or maybe a suggestion for a different framework?

**Reply by python-and-novella** [reply]
Maybe Qt,it is a GUI framework which is more compatible with threading.

**Comment by rodja** [comment, marked as answer]
It heavily depends on what the usage is. With web UI frameworks you generally have a multi-user setup: each browser tab is a user, doing its interactions and navigation between pages. With https://rosys.io we made a robotics framework based on NiceGUI where multiple users can access and interact simultaneously with the same hardware. I guess something similar is applying in your case where you already have a large codebase and state transitions between them happen independently form what the user is doing. A way to do this from a background thread would be like this:

```py
import asyncio
from nicegui import app, ui, Client, background_tasks

event = asyncio.Event()

@ui.page("/")
def main_page():
    ui.label("start_page")
    background_tasks.create(check_state(ui.context.client))

@ui.page("/pagetwo")
def second_page():
    ui.label("Second page")

def threaded_task():
    time.sleep(10)
    event.set()

async def check_state(client: Client):
    await event.wait()
    with client:
        ui.navigate.to("/pagetwo")

task = threading.Thread(target=threaded_task)
task.start()

ui.run()
```

There are of course other ways to signal state changes from a thread to an asyncio event loop. But `asyncio.Event` is a simple solution. You could also simply replace some ui elements, use a `ui.refreshable` or anything else to reflect the state of your app. For example:

```py
import asyncio
from nicegui import ui

event = asyncio.Event()

@ui.page("/")
async def main_page():
    await ui.context.client.connected()
    with ui.column() as status:
        ui.label("start")
    await event.wait()
    status.clear()
    with status:
        ui.label("finish")

def threaded_task():
    time.sleep(3)
    event.set()

task = threading.Thread(target=threaded_task)
task.start()
ui.run()
```

**Reply by NWHBakker** [reply]
Works like a charm, Thanks a lot.
The rosys implementation might be an interesting starting point the next time. We have been avoiding ROS for this project. 
In hindsight it might have been a good solution.
